### PR TITLE
Update Code Coverage Tools & Documentation

### PR DESCRIPTION
**Pull request by dmacattack**
- remove coverage script that relies on nightly builds
- rename script using tarpaulin coverage
- add coverage script using llvm-cov
- update ignore
- add readme for tools

example report - extract and open html/index.html
[up-streamer-rust-llvm-coverage-example.zip](https://github.com/user-attachments/files/16639915/up-streamer-rust-llvm-coverage-example.zip)

### CODE REVIEW

File: tools/coverage.sh [hunk 1]
```diff
@@ -1,21 +0,0 @@
...
-
-PROJECT_NAME_UNDERSCORE="up_streamer_rust"
-RUSTFLAGS="--cfg uuid_unstable -Zprofile -Ccodegen-units=1 -Copt-level=0 -Clink-dead-code -Coverflow-checks=off -Zpanic_abort_tests -Cpanic=abort"
-RUSTDOCFLAGS="-Cpanic=abort"
-TEMP=`mktemp --directory`
-
-
-mkdir -p $TEMP/reports
-
-cargo test $CARGO_OPTIONS -- -Z unstable-options --format json | cargo2junit > $TEMP/results.xml
```
Comment (dmacattack) [inline tools/coverage.sh:14]: -Z and following args have dependency on nightly builds, which makes the rest of the script unusable/not work

File: tools/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Generate coverage using the llvm coverage tool. The output files are generated to reports/llvm/html/index.html, reports/llvm/lcov.info & reports/llvm/lcov/index.html
+
+``` bash
+$ sh tools/generate_llvm_coverage.sh 
+... 
+Writing directory view page.
+Overall coverage rate:
+  lines......: 78.1% (890 of 1140 lines)
+  functions..: 55.1% (201 of 365 functions)
```
Comment (PLeVasseur) [inline tools/README.md:15]: Wowzers, there's a lot more functions captured by llvm coverage. Can you attach a sample report of each to this PR and maybe speak to some of the differences in functions inspected in the PR summary?

File: tools/generate_llvm_coverage.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+if [ -z "$HAS_LLVMCOV" ]; then
+    echo "cargo-llvm-cov not found, please install it with 'cargo install cargo-llvm-cov'"
+    exit 1
+fi
+
+mkdir -p $COVERAGE_OUT
+
+# generate coverage using llvm
+# RUST_LOG ensures the trace macros are not flagged by the coverage output
```
Comment (PLeVasseur) [inline tools/generate_llvm_coverage.sh:16]: So the `trace!()` macros were the only issue with code coverage? No other log macros?

File: tools/generate_llvm_coverage.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+BLUE='\033[0;34m'
+NC='\033[0m'
+HAS_LLVMCOV=$(cargo install --list | grep cargo-llvm-cov)
```
Comment (PLeVasseur) [inline tools/generate_llvm_coverage.sh:5]: :+1:

File: tools/generate_llvm_coverage.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
```
Comment (PLeVasseur) [inline tools/generate_llvm_coverage.sh:2]: Please add copyright header as in `generate_tarpaulin_coverage.sh`

**Review by PLeVasseur — COMMENTED** [review]
Hey @dmacattack -- take a look at some requests

File: tools/generate_llvm_coverage.sh (hunk 3 above)
Comment (dmacattack) [reply tools/generate_llvm_coverage.sh:16]: trace is the log level. this "trips" all levels, and has favorable coverage. I suppose i should remove the 'trace' from the comment to be more accurate

File: tools/generate_llvm_coverage.sh (hunk 5 above)
Comment (dmacattack) [reply tools/generate_llvm_coverage.sh:2]: will do

File: tools/README.md (hunk 2 above)
Comment (dmacattack) [reply tools/README.md:15]: we did notice a disparity between the tools, and some of the coverage has to do with _how_ the coverage is highlighted. Also the TRACE/DEBUG/WARN macros were not ignored in tarpaulin as they are with llvm - which could be the line coverage. 

Like any tool, the coverage needs an educated dev to assess the results given.

Comment (PLeVasseur) [reply tools/README.md:15]: Thanks for updating with the report. I see entries that are captured for `utils/`:

image

Could you check for a way to exclude that folder? It contains only utils code for making the examples and integration test infra.

**Review by PLeVasseur — CHANGES_REQUESTED** [review]
Hey @dmacattack -- left a request, could you take a look?

File: tools/README.md (hunk 2 above)
Comment (dmacattack) [reply tools/README.md:15]: yes good call, i did it in tarpaulin. it should be available in llvm

**Comment by dmacattack** [comment]
> Hey @dmacattack -- left a request, could you take a look?

completed

File: tools/README.md (hunk 2 above)
Comment (dmacattack) [reply tools/README.md:15]: image
 no more utils

Comment (PLeVasseur) [reply tools/README.md:15]: All right, looks good!

**Review by PLeVasseur — APPROVED** [review]
Thanks @dmacattack, LGTM!

Can you now clean up the commit history to focus on the feature, perhaps a single squashed commit?

Please also use the format:

```
[#XX] PR name

Implements [#YY]
```

where #XX is the PR # and #YY is the issue this is solving?

**Comment by dmacattack** [comment]
> [#XX] PR name

@PLeVasseur done & done

**Pull request merged**